src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs

```diff
@@ -771,30 +772,202 @@ public class B
+
+            var field1 = (IFieldSymbol)((INamedTypeSymbol)model.GetDeclaredSymbol(member1Syntax)).GetMembers("field").Single(); // A<T!>! A<T>.field
+            var field2 = (IFieldSymbol)model.GetSymbolInfo(member2Syntax).Symbol;                                               // A<T!>! A<T!>.field
+
+            VerifyEquality(field1, field2,
+                expectedDefault: true,
+                expectedIncludeNullability: false
+                );
+
+            var field1Type = field1.Type; // A<T!>
+            var field2Type = field2.Type; // A<T!>
+
+            VerifyEquality(field1Type, field2Type,
+                expectedDefault: true,
+                expectedIncludeNullability: true
+                );
+
+            var field1ContainingType = field1.ContainingType; //A<T>
+            var field2ContainingType = field2.ContainingType; //A<T!>
+
+            VerifyEquality(field1ContainingType, field2ContainingType,
+                expectedDefault: true,
+                expectedIncludeNullability: false
+                );
+
+        }
+
+        [Fact]
+        [WorkItem(8195, "https://github.com/dotnet/roslyn/issues/38195")]
+        public void SemanticModel_SubstitutedMethod_Equality()
+        {
+            var source =
+@"
+#nullable enable
+public class A<T> 
+    where T : class //not neccesary, but makes it easier to reason about the resulting fields
+{
+    public A<T> M(A<T> t)
+    {
+        t.M(t);
+        return t;
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+
+            var syntaxTree = comp.SyntaxTrees[0];
+            var root = syntaxTree.GetRoot();
+
+            var member1Syntax = (ClassDeclarationSyntax)root.DescendantNodes().First(sn => sn.Kind() == SyntaxKind.ClassDeclaration);
+            var member2Syntax = (IdentifierNameSyntax)root.DescendantNodes().Last(sn => sn.Kind() == SyntaxKind.SimpleMemberAccessExpression).DescendantNodes().Last(sn => sn.Kind() == SyntaxKind.IdentifierName);
+
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var method1 = (IMethodSymbol)((INamedTypeSymbol)model.GetDeclaredSymbol(member1Syntax)).GetMembers("M").Single(); // A<T!>! A<T>.M(A<T!>! t)
+            var method2 = (IMethodSymbol)model.GetSymbolInfo(member2Syntax).Symbol;                                           // A<T!>! A<T!>.M(A<T!>! t)
+
+            VerifyEquality(method1, method2,
+                expectedDefault: true,
+                expectedIncludeNullability: false
+                );
+
+            var method1ReturnType = method1.ReturnType; // A<T!>
+            var method2ReturnType = method2.ReturnType; // A<T!>
+
+            VerifyEquality(method1ReturnType, method2ReturnType,
+                expectedDefault: true,
+                expectedIncludeNullability: true
+                );
+
+            var method1ParamType = method1.Parameters.First().Type; // A<T!>
```

Comment:
Please also test the `ParameterSymbol`s directly.

Reply:
Can you think of a way of getting different parameter symbols, such that one is A<T> and the other A<T!>? I tried to come up with a way, but couldn't.

Do you want me to just break it out and test as is even if we cant?

---
In reply to: [366075635](https://github.com/dotnet/roslyn/pull/40939#discussion_r366075635) [](ancestors = 366075635)